src/System.IO.FileSystem/src/System/IO/Win32FileStream.cs

```diff
@@ -1291,19 +1305,45 @@ private Task WriteInternalAsync(byte[] array, int offset, int numBytes, Cancella
                 _readLen = 0;
             }
 
-            int n = _bufferSize - _writePos;
-            if (numBytes <= n)
+            int remainingBuffer = _bufferSize - _writePos;
+
+            // if the incoming write fits in the remaining buffer, treat
+            // as synchronous since memcopy is so fast
+            if (numBytes <= remainingBuffer)
             {
                 if (_writePos == 0) _buffer = new byte[_bufferSize];
                 Buffer.BlockCopy(array, offset, _buffer, _writePos, numBytes);
                 _writePos += numBytes;
-                
+
                 return Task.CompletedTask;
             }
 
-            if (_writePos > 0)
-                FlushWrite(false);
+            // if the incoming write fits in the current buffer + next buffer,
+            // only perform a single write to flush the existing buffer and fill
+            // the next buffer with the remainder
+            if (numBytes <= (_bufferSize + remainingBuffer))
+            {
+                if (_writePos == 0) _buffer = new byte[_bufferSize];
```

Comment:
It depends on what guarantees we make about using the async APIs on FileStream.  If we say that it's valid to do:
```C#
Task t1 = fileStream.WriteAsync(data, 0, data.Length);
Task t2 = fileStream.WriteAsync(data, 0, data.Length);
await Task.WhenAny(t1, t2);
```
then, yeah, that would explain it.  Many stream say that's not valid, and either throw if you try to do that or serialize the calls rather than try to overlap them.  @ericstj, what guarantees do we make about FileStream in this regard?